File: backend/app/schemas/compras.py
from datetime import date
from typing import Annotated

from pydantic import BaseModel, Field, field_validator


class CompraItemCreate(BaseModel):
    descricao_original: Annotated[str, Field(min_length=1, max_length=240)]
    quantidade: float = Field(gt=0, le=1_000_000)
    unidade: Annotated[str, Field(min_length=1, max_length=20)] = "un"
    valor_unitario: float = Field(ge=0, le=100_000_000)
    valor_total: float = Field(ge=0, le=100_000_000)

    @field_validator("descricao_original", "unidade")
    @classmethod
    def strip_text(cls, value: str) -> str:
        return " ".join(value.split())


class CompraNfceCreate(BaseModel):
    qr_texto: Annotated[str, Field(min_length=8, max_length=4_000)]
    chave_nfce: Annotated[str, Field(max_length=60)] = ""
    mercado_nome: Annotated[str, Field(min_length=2, max_length=160)]
    cnpj: Annotated[str, Field(max_length=30)] = ""
    data_compra: date
    valor_total: float = Field(gt=0, le=100_000_000)
    forma_pagamento: Annotated[str, Field(max_length=100)] = ""
    valor_pago: float = Field(ge=0, le=100_000_000)
    itens: list[CompraItemCreate] = Field(min_length=1, max_length=500)

    @field_validator("chave_nfce")
    @classmethod
    def normalize_key(cls, value: str) -> str:
        digits = "".join(character for character in value if character.isdigit())
        return digits if len(digits) == 44 else ""

    @field_validator("mercado_nome", "forma_pagamento", "cnpj")
    @classmethod
    def strip_fields(cls, value: str) -> str:
        return " ".join(value.split())


class CompraCreateResponse(BaseModel):
    compra_id: str
    familia_id: str
    supermercado_id: str
    itens_salvos: int = Field(ge=1)
    produtos_criados: int = Field(ge=0)
    produtos_reutilizados: int = Field(ge=0)
    mensagem: str

File: frontend/types/compras.ts
export type CompraSalva = {
  compra_id: string;
  familia_id: string;
  supermercado_id: string;
  itens_salvos: number;
  produtos_criados: number;
  produtos_reutilizados: number;
  mensagem: string;
};

export type CompraExcluida = {
  compra_id: string;
  familia_id: string;
  itens_excluidos: number;
  historicos_excluidos: number;
  mensagem: string;
};

export type CompraResumo = {
  id: string;
  supermercado_nome: string;
  data_compra: string;
  valor_total: number;
  forma_pagamento: string;
  status: string;
  itens_count: number;
  criado_em: string;
};

export type CompraLista = {
  compras: CompraResumo[];
  limite: number;
  offset: number;
  proximo_offset: number | null;
  tem_mais: boolean;
};

export type CompraItemDetalhe = {
  id: string;
  produto_id: string | null;
  produto_nome: string;
  descricao_original: string;
  quantidade: number;
  unidade: string;
  valor_unitario: number;
  valor_total: number;
  categoria_nome: string;
};

export type CompraDetalhe = {
  id: string;
  familia_id: string;
  supermercado_id: string | null;
  supermercado_nome: string;
  supermercado_cnpj: string;
  chave_nfce: string;
  data_compra: string;
  valor_total: number;
  forma_pagamento: string;
  valor_pago: number;
  origem: string;
  status: string;
  criado_em: string;
  itens: CompraItemDetalhe[];
};
